aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -0,0 +1,326 @@
+  gradWeight[weightRow + startFeature] = weight;
+}
+
+} // anon namespace
+
+Tensor embedding_backward_cuda_kernel(
+        const Tensor &grad,
+        const Tensor &orig_indices,
+        const Tensor &sorted_indices,
+        const Tensor &count,
+        int64_t num_weights,
+        int padding_idx,
+        bool scale_grad_by_freq,
+        bool mode_mean,
+        const Tensor &offset2bag,
+        const Tensor &bag_size,
+        const Tensor &per_sample_weights) {
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  const ptrdiff_t numel = sorted_indices.numel();
+
+  auto grad_weight = at::zeros({num_weights, grad.size(-1)}, grad.options());
+  const int64_t stride = grad_weight.stride(0);
+
+  // Compute the number of segments and their start position so that we do not have to
+  // spawn a warp per index. In this context, a segment is a number of rows that should
+  // be summarized.
+  // Unit: index in `sorted_indices` and `orig_indices`
+  thrust::device_vector<int64_t> segment_offsets(numel);
+  int64_t num_of_segments;
+  {
+    auto sorted_indices_dev = thrust::device_ptr<int64_t>(sorted_indices.data<int64_t>());
+    auto dummy = at::empty_like(sorted_indices);
+    auto dummy_dev = thrust::device_ptr<int64_t>(dummy.data<int64_t>());
+    auto ends = thrust::unique_by_key_copy(
+            policy,
+            sorted_indices_dev,
+            sorted_indices_dev + numel,
+            thrust::make_counting_iterator(0),
+            dummy_dev,
+            thrust::raw_pointer_cast(segment_offsets.data()));
+    num_of_segments = thrust::get<0>(ends) - dummy_dev;
+  }
+
+  // We split the segments up into sizes of `NROWS_PER_THREAD`
+  // Compute the number partial-segments per segment (some partial-segments 
+  // may not be the full `NROWS_PER_THREAD` number of rows)
+  thrust::device_vector<int64_t> partials_per_segment(num_of_segments);
+  {
+    krn_partials_per_segment<<<ceil_div(num_of_segments, 32), 32, 0, stream>>> (
+            thrust::raw_pointer_cast(partials_per_segment.data()),
+            thrust::raw_pointer_cast(segment_offsets.data()),
+            num_of_segments,
+            numel);
+  }
+
+  // In order to compute `partial_segment_offset`, which is the start index
+  // of each partial-segment in `sorted_indices`, we need to compute the
+  // start position of each _segment_ in `partial_segment_offset`.
+  // Unit: index in `partial_segment_offset`
+  thrust::device_vector<int64_t> partials_per_segment_offset(num_of_segments); 
+  thrust::exclusive_scan(
+          policy,
+          partials_per_segment.begin(),
+          partials_per_segment.end(),
+          partials_per_segment_offset.begin());
+
+  // The total number of partial-segments is the sum of `partials_per_segment_offset`
+  const int num_of_partial_segments = partials_per_segment[num_of_segments-1] +
```

Comment:
Direct accesses to device_vector result in cudaMemcopy, synchronizing no less. They should be used very sparingly, if there's literally no way around it.